### DISCUSSION TITLE
Depth of Field: incorrect masking with transparent materials

### DISCUSSION THREAD

**Discussion by davcri** [pmndrs/postprocessing#610] [Q&A]
### Description of the bug

Rendered image isn't correct when using DoF with transparent materials.

### To reproduce

https://codesandbox.io/p/sandbox/react-postprocessing-dof-isssue-forked-2zcyyg

### Screenshots

_UPDATE: the original post contained only the previous sandbox. I added these screenshots afterwards._ 

<img width=400 src="https://github.com/pmndrs/postprocessing/assets/6860637/546bbf0d-880c-4e15-b301-d7e94b990caf" />

[Codesandbox link](https://codesandbox.io/p/sandbox/weathered-sun-l6c7km?file=%2Fsrc%2Findex.js%3A38%2C18-38%2C23): Current rendering with 6.34.3. Cube with transparent MeshStandardMaterial and envmap reflections.

### Desired effect

<img width=400 src="https://github.com/pmndrs/postprocessing/assets/6860637/d1cb6148-a671-45b3-9a6f-56b73a3c7745" />

[Codesandbox link](https://codesandbox.io/p/devbox/threejs-pp-forked-nqvlph?file=%2Fsrc%2Findex.js%3A154%2C1): custom postprocessing to illustrate the desired rendering with transparent MeshStandardMaterial, sharp edges and blurred background. The code does not take depth into account and it probably has wrong light calculations.
There is also a r3f version [here](https://codesandbox.io/p/sandbox/youthful-wave-njccwt?file=%2Fsrc%2FPostprocessing.jsx).

### Workarounds

- `depthWrite=false`:  fixes the background but blurs also the edges of the transparent object.

<img width=400 src="https://github.com/pmndrs/postprocessing/assets/6860637/ba5f69d9-aeb7-4c3d-a8bc-6cada4c3e874" />

**Comment by vanruesc** [comment]
The cube is marked as transparent but its opacity is 1. To me this looks like it's working as intended: the reflective cube is in focus and the background is blurred.

**Comment by davcri** [comment]
The example is a bit misleading since I'm using a `MeshPhysicalMaterial` with transmission (`transparency=true` has no effect).

However I get the same result by using a `MeshStandardMaterial` with transparency and opacity < 1.0 (https://codesandbox.io/p/sandbox/react-postprocessing-dof-isssue-601-meshstandardmaterial-p4grw6?file=%2Fsrc%2FApp.js%3A65%2C74)

---

The main issue is that the object writes to the depth buffer and this avoids blurring the background, even if it should be.
![image](https://private-user-images.githubusercontent.com/6860637/306030573-65cc514b-c9a5-424f-80be-ced17499f128.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA4MzgzNTgsIm5iZiI6MTcxMDgzODA1OCwicGF0aCI6Ii82ODYwNjM3LzMwNjAzMDU3My02NWNjNTE0Yi1jOWE1LTQyNGYtODBiZS1jZWQxNzQ5OWYxMjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTlUMDg0NzM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWZkYWVjMmQ3ZmQ3ODJlZmE5NDMyMjIyNmE0ZmVkY2NlODhkMDYwYmQ5ZDBjZjc1YzdhYjMwNGUxOTgwN2UzMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.-5--p4hfRlITUe-Gu8-FdMnd72oHBBEYWqqE-P2gWzI)

Another workaround would be increasing `roughness` on the MeshPhysicalMaterial, however I think this isn't still a realistic effect (since reflections would be affected, in a similar way to `depthWrite=false`).

**Comment by vanruesc** [comment]
This won't work since the `DepthOfFieldEffect` has no influence on the textures that are used for transmission/reflection sampling. Does [backgroundBlurriness](https://threejs.org/docs/?q=scene#api/en/scenes/Scene.backgroundBlurriness) address your use case?

**Comment by davcri** [comment]
It would work for envmap backgrounds but it wouldn't work with proper meshes (which unfortunately is what i have in the project I was working on while I found out the bug).

I took this image from a [Unity thread](https://forum.unity.com/threads/transparent-objects-not-working-with-dof.618184/), hopefully this clarifies what I'd like to achieve:

![image](https://private-user-images.githubusercontent.com/6860637/306039253-923c802a-1b2d-493e-8bbb-10d8ca6ec7ef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA4MzgzNTgsIm5iZiI6MTcxMDgzODA1OCwicGF0aCI6Ii82ODYwNjM3LzMwNjAzOTI1My05MjNjODAyYS0xYjJkLTQ5M2UtOGJiYi0xMGQ4Y2E2ZWM3ZWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTlUMDg0NzM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWI1MWU0ZWYxNjAzNWIyZTBiMDllODk4YzBjY2Y1MTFlOTU4ZjU4YTRmMmM2MDQ3MjJkMTBlMjI1MzFkZmEwYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.obfBNB3ozAlimNV88q7kvZXVh2dCr8l__2oXb7CEO_k)

PS: thanks again for all the replies @vanruesc , sorry for being pedantic but I didn't find anything online regarding ThreeJS and I'd like to find a fix for this (probably not so common) problem.

**Comment by vanruesc** [comment]
> The main issue is that the object writes to the depth buffer and this avoids blurring the background, even if it should be.

The `DepthOfFieldEffect` blurs objects based on depth and that's working as intended in this case. We can't see the actual background through the transmissive object - that's effectively just its texture.

The deeper problem with transmission is that the renderer [renders the scene to a transmission texture](https://github.com/mrdoob/three.js/blob/cfdd878cf74857aef26459eb82c68dd75f097c72/src/renderers/WebGLRenderer.js#L1430) in an internal process that can't be modified. If this was customizable, we might be able to apply DoF to the transmission texture that is later read by the transmissive object. We'd then have to render DoF again for the fully rendered scene, but we'd end up blurring the transmissive object twice...

I don't know if there's a solution to this.

**Comment by davcri** [comment]
Yeah you're right about MeshPhysicalMaterial, however the problem affects also other materials with transparency, correct? 

 > I don't know if there's a solution to this.
 
Me neither, but I'd lke to investigate about the required changes in ThreeJS and Postprocessing in order to have a proper support for DoF and transparent objects out of the box.
Do you know anyone in the community that could give some hint?

---

In the meantime I updated the first comment with:
1. a simplified explanation of the problem/workarounds
2. a [codesanbox with a simulation](https://codesandbox.io/p/devbox/threejs-pp-forked-nqvlph?file=%2Fsrc%2Findex.js%3A154%2C1) of the desired effect (naive approach that doesn't take depth into account)

In the simulation I did something like this:
```js
composer.addPass(new RenderPass(sceneWithOpaqueObjects, camera));
composer.addPass(new EffectPass(camera, new DepthOfFieldEffect(camera, {...})));
composer.addPass(new CopyPass(opaqueRenderTarget));
composer.addPass(new ClearPass(true, true));
composer.addPass(new RenderPass(sceneWithTransparentCube, camera));
composer.addPass(new CopyPass(transparentRenderTarget));
// render transparent texture on top of the opaque texture
composer.addPass(new EffectPass(camera, new CustomEffect(transparentRenderTarget.texture, opaqueRenderTarget.texture)));
```

**Comment by vanruesc** [comment]
> however the problem affects also other materials with transparency, correct?

Yes, it applies to all transparent meshes. However, transmission is a different problem compared to true transparency. And transparency comes with different blend modes: alpha-blend, alpha-test and alpha-hash. 
- Alpha-Blend has the same issue as you've encountered with transmission.
- Alpha-test is not affected because it's just a cutout based on alpha.
- Alpha-hash with TAA might just work, but I haven't tried that since we don't have a good `TAAEffect` yet.

> Do you know anyone in the community that could give some hint?

No, and I don't think anyone would have a definitive answer. Probably relevant: https://forum.unity.com/threads/transparent-rendering-with-dof.784823/

**Comment by hybridherbst** [comment]
> proper support for DoF and transparent objects out of the box

That one's a hard problem as there's only one layer of depth buffer (in typical implementations) and it's not clear where the DoF on transparent objects should "go to". E.g. at opacity 0.1 you likely want blurred background (doesn't write depth), and at 0.9 you likely want no blur (writes depth) for a focus point on the transparent object.

Transmission has the same issue in the other direction (it writes depth, and at transmission = 0.1 you
d expect it to behave like any other opaque object, but at transmission=1.0 you'd expect behaviour closer to transparent).

There are workarounds as mentioned above:
- expensively compositing multiple layers (what movie production does)
- various dithering approaches ("alphaHash" in three.js) that can work ok for full-resolution depth buffers depending on the DoF implementation

**Comment by vanruesc** [comment]
I feel like this is one of those issues that are bound to stay open forever since there's no clear solution. Would it be ok if I convert it into a discussion instead?

**Comment by davcri** [comment]
@vanruesc ok for me

**Comment by vanruesc** [comment]
We can check if alpha-hash + TAA is a viable option once `TAAEffect` is implemented.
